### ISSUE TITLE
dynamic batching does not work in decoupled model

### ISSUE THREAD

**Issue by scse-l** [triton-inference-server/server#5926]
**Description**
I deployed a decoupled model with python backend. During tests, dynamic batching does not work as expected.
As documented in https://github.com/triton-inference-server/server/blob/main/docs/user_guide/decoupled_models.md:
> C++ Backend
... This can also cause under-utilization of features like [Dynamic Batching](https://github.com/triton-inference-server/server/blob/main/docs/user_guide/model_configuration.md#dynamic-batcher) as it leads to eager batching.

So does it lead to eager batching too in decoupled models with python backend ? 

**Triton Information**
What version of Triton are you using?
22.10

Are you using the Triton container or did you build it yourself?
Triton container

**To Reproduce**
Steps to reproduce the behavior.

Describe the models (framework, inputs, outputs), ideally include the model configuration file (if using an ensemble include the model configuration file for that as well).

**Expected behavior**

**Comment by nv-kmcgill53** [comment]
Hi @scse-l, I see you are using an older version of triton, 22.10. As a quick test, can you see if there is still an issue in the latest release, 23.05?  Also, to help us, could you fill out the other fields of your report, specifically "To Reproduce" and "Expected Behavior?" If you are not able to share the python model which you are using, are you able to provide a minimal example to us? Thanks.

**Comment by rmccorm4** [comment]
Hi @scse-l,

The python backend is unique compared to some of the other backends (TF, PyTorch, etc.) in that it doesn't currently implement the actual "batching" logic for you, which in simpler terms is:
1. Collecting the inputs from multiple compatible requests into one larger batch
2. Executing the one large batch
3. Scattering the batch back out into individual responses for each request

It supports dynamic batching in the sense that Triton can gather the requests in the server/core based on the dynamic batching settings for the model, and then send those requests together in a single API call to the model. Even without the logic above, this should provide marginal benefit in that it requires less communication (API calls) with the model. However, the larger benefit is in the (2) described above, which is computing the single large batch rather than computing multiple smaller batches. 

This is something that can you can implement in your python model, but the range of things you can do with python models is vast, so it has not been extracted out into some common python backend utility or something that is done under the hood for you today like other backends. 

Here is some **very rough pseudocode** to express the general idea, assuming a model with a single input called "INPUT0", but can be extended:
```python
def initialize(self, args):
  # ...
  self.model = MyModel()

def execute(self, requests):
  large_input = []
  all_inputs = []
  all_batch_sizes = []
  for request in requests:
    np_input = pb_utils.get_input_tensor_by_name(request, "INPUT0").as_numpy()
    batch_size = np_input.shape[0]
    all_inputs.append(np_input)
  
  # Gather the request inputs into a single batch
  batched_input = np.stack(all_inputs, axis=0)

  # Compute large batch all together into single large output batch
  batched_output = self.model.infer(batched_input)
  
  # Python model execute() must return a response for each request received,
  # so split up the batched_output into individual response outputs for each request batch size
  responses = []
  for i, batch_size in enumerate(all_batch_sizes):
    # Break the large output back into individual outputs for each request batch size
    individual_output = get_batch(batched_output, batch_size, i)
    # Create output tensors mapping each request to each response for their respective batch sizes
    # ...
    responses.append(...)

  assert len(responses) == len(requests)
  return responses
```

The idea here is that we only did a single expensive computation on all the requests at once (`self.model.infer(batched_input)`), and then break out the large output to create individual responses for each request for their respective batch sizes.

This depends on your python model's internal logic actually supporting the idea of executing in batches.

Hope this makes sense.

**Comment by scse-l** [comment]
> Hi @scse-l,
> 
> The python backend is unique compared to some of the other backends (TF, PyTorch, etc.) in that it doesn't currently implement the actual "batching" logic for you, which in simpler terms is:
> 
> 1. Collecting the inputs from multiple compatible requests into one larger batch
> 2. Executing the one large batch
> 3. Scattering the batch back out into individual responses for each request
> 
> It supports dynamic batching in the sense that Triton can gather the requests in the server/core based on the dynamic batching settings for the model, and then send those requests together in a single API call to the model. Even without the logic above, this should some provide marginal benefit in that it requires less communication (API calls) with the model. Internally this is like Triton saying `model.execute([request1, request2])` instead of `model.execute(request1); model.execute([request2])`
> 
> However, the largest benefit is likely in step (2) described above, which is computing the single large batch rather than computing multiple smaller batches. This is something that can you can implement in your python model, but the range of things you can do with python models is vast, so it has not been extracted out into some common python backend utility or something that is done under the hood for you today like other backends.
> 
> Here is some **very rough pseudocode** to express the general idea, assuming a model with a single input called "INPUT0", but can be extended:
> 
> ```python
> def initialize(self, args):
>   # ...
>   self.model = MyModel()
> 
> def execute(self, requests):
>   large_input = []
>   all_inputs = []
>   all_batch_sizes = []
>   for request in requests:
>     np_input = pb_utils.get_input_tensor_by_name(request, "INPUT0").as_numpy()
>     batch_size = np_input.shape[0]
>     all_inputs.append(np_input)
>   
>   # Gather the request inputs into a single batch
>   batched_input = np.stack(all_inputs, axis=0)
> 
>   # Compute large batch all together into single large output batch
>   batched_output = self.model.infer(batched_input)
>   
>   # Python model execute() must return a response for each request received,
>   # so split up the batched_output into individual response outputs for each request batch size
>   responses = []
>   for i, batch_size in enumerate(all_batch_sizes):
>     # Break the large output back into individual outputs for each request batch size
>     individual_output = get_batch(batched_output, batch_size, i)
>     # Create output tensors mapping each request to each response for their respective batch sizes
>     # ...
>     responses.append(...)
> 
>   assert len(responses) == len(requests)
>   return responses
> ```
> 
> The idea here is that we only did a single expensive computation on all the requests at once (`self.model.infer(batched_input)`), and then break out the large output to create individual responses for each request for their respective batch sizes.
> 
> This depends on your python model's internal logic actually supporting the idea of executing in batches.
> 
> Hope this makes sense.

The internal logic is almost the same as the sample you shows.
```python
    def execute(self, requests):
        input_list = []
        batch_size = len(requests)
        sender_list = []
        for request in requests:
            input_list.append(pb_utils.get_input_tensor_by_name(request, 'input_name').as_numpy())
            sender_list.append(request.get_response_sender())

        gen_thread = Thread(target=self.model.inference, some args...)
        gen_thread.setDaemon(True)
        gen_thread.start()

        [generate stream responses from response iterators]
        ...
```
However, the batch_size is always 1.

**Comment by scse-l** [comment]
@rmccorm4 
OK, I will try it in 23.05.
